signer/contentsignaturepki/upload_test.go

```diff
@@ -0,0 +1,76 @@
+package contentsignaturepki
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"testing"
+
+	"github.com/aws/aws-sdk-go-v2/feature/s3/manager"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+)
+
+// Mocks adapted from https://aws.github.io/aws-sdk-go-v2/docs/unit-testing/
+type mockUploadAPI func(ctx context.Context, input *s3.PutObjectInput, opts ...func(*manager.Uploader)) (*manager.UploadOutput, error)
+
+func (m mockUploadAPI) Upload(ctx context.Context, input *s3.PutObjectInput, opts ...func(*manager.Uploader)) (*manager.UploadOutput, error) {
+	return m(ctx, input, opts...)
+}
+
+func TestUploadToS3(t *testing.T) {
+	cases := []struct {
+		client    func(t *testing.T) S3UploadAPI
+		data      string
+		name      string
+		target    string
+		expectErr bool
+	}{
+		{
+			client: func(t *testing.T) S3UploadAPI {
+				return mockUploadAPI(func(ctx context.Context, input *s3.PutObjectInput, opts ...func(*manager.Uploader)) (*manager.UploadOutput, error) {
+					t.Helper()
+					return &manager.UploadOutput{}, nil
+				})
+			},
+			data:      "foo",
+			name:      "successful_upload",
+			target:    "https://foo.bar",
+			expectErr: false,
+		},
+		{
+			client: func(t *testing.T) S3UploadAPI {
+				return mockUploadAPI(func(ctx context.Context, input *s3.PutObjectInput, opts ...func(*manager.Uploader)) (*manager.UploadOutput, error) {
+					t.Helper()
+					return nil, errors.New("upload failed")
+				})
+			},
+			data:      "foo",
+			name:      "failed_upload",
+			target:    "https://foo.bar",
+			expectErr: true,
+		},
+	}
+
+	for _, tt := range cases {
+		t.Run(tt.name, func(t *testing.T) {
+			tt := tt
```

Comment:
Oh, [Go 1.22 changed how for-loops work](https://go.dev/blog/loopvar-preview) so we no longer need to worry about capturing like this. 

But if we wanted to continue doing that for now as practice for the Go code we rely on that's not yet on Go 1.22, I think you want this redefinition outside of the closure to avoid racing on it (otherwise, the capture has already happened). So move it above the `t.Run` call.